app/utils/ServerRequests.tsx

```diff
@@ -292,13 +314,269 @@ export const uploadFile = async (file: Blob, location: string ): Promise<boolean
         body: raw,
         redirect: "follow"
     };
-    
-    return fetch(machineURL+location, requestOptions as RequestInit)
-        .then((response) => response.text())
-        .then((result) => { return true })
+
+    return fetch(machineURL + location, requestOptions as RequestInit)
+        .then((response) => { return response.text() })
+        .then(async (result) => {
+            let fileID: number
+            await getFolderContent(location).then(async (response: void | FileListType) => {
+                if (response) {
+                    fileID = response.files[response.files.length - 1].fileID;
+                    let tags = await getAllSystemTags().then((tags: void | FileTagType[]) => {
+                        let file_name = response.files[response.files.length - 1].fileName.toLowerCase();
+                        let directory_path = location; // assuming directory_path is same as location
+
+                        if (!tags) {
+                            console.error("No tags found in system");
+                            return false;
+                        }
+
+                        // Assign tags based on directory path
+                        ExampleTags['Subjects'].forEach((tag) => {
+                            if (directory_path.includes(tag)) {
+                                tags.forEach(async (systemTag: any) => {
+                                    if (systemTag.tagName === tag) {
+                                        if (!(await assignSystemTag(fileID, systemTag))) {
+                                            console.error("Failed to assign tag in Subject.");
+                                        }
+                                    }
+                                });
+                            }
+                        });
+                        
+                        // Assign user tag to all uploaded files
+                        tags.forEach(async (systemTag) => {
+                            if (systemTag.tagName === user) {
+                                if (!(await assignSystemTag(fileID, systemTag))) {
+                                    console.error("Failed to assign User Tag.");
+                                }
```

Comment:
Thsi ``forEach`` you used everywhere doesn't seem necessary. Simply index the systemTag by it's name. The data structure could look like:
```ts
export interface FileTagType {
    [tagName: string]: string
}
```
So the tagName is the key and the value would be the id. I think this should work.

Reply:
We talked about it and we will do the refactor later because other issues depend on this MR and we have time pressure.